### ISSUE TITLE
[feature request] Pre-training, GPL, LLM2Vec

### ISSUE THREAD

**Issue by dmacko232** [UKPLab/sentence-transformers#2689]
Hello @tomaarsen,

I hope u are doing well. I love the new Sentence Transfomers release!

I am writing to propose a few feature requests that I believe would enhance Sentence Transformers:
- Add examples of more pre-training approaches such as [RetroMAE and DupMAE](https://github.com/staoxiao/RetroMAE) which have shown SOTA performance when pre-training for retrieval. It would be nice to be able to easily use them in Sentence Transformers without relying on another library.
-  Example of fine-tuning using [GPL](https://github.com/UKPLab/gpl), as the training code changed and it could be done more efficiently, compactly, and directly withing Sentence Transformers instead of relying on another library. Additionally, this could enable pseudo-labeling using LLMs, such as GPT-4, instead of Cross-encoders. It would be also nice to decompose this in a way that it is possible to easily generate queries (GenQ) and mine hard negatives without the pseudo-labeling part (in case the pseudo-labeling is not desired).
- Add easy merging of fine-tuned and base model, inspired from [LM-Cocktail](https://github.com/FlagOpen/FlagEmbedding/tree/master/LM_Cocktail)
- Integration with [LLM2Vec](https://github.com/McGill-NLP/llm2vec)

**Comment by austinmw** [comment]
Would also like to see these features!

**Comment by tomaarsen** [comment]
Hello!

- [ ] I completely agree. RetroMAE/DupMAE are quite valuable and commonly used in research. They use a wholly different training approach than Sentence Transformers, but I think some examples & documentation on RetroMAE/DupMAE would be useful, e.g. very much like this [MLM example](https://github.com/UKPLab/sentence-transformers/blob/master/examples/unsupervised_learning/MLM/train_mlm.py). I'm definitely open to PRs here!
- [ ] I'm aware of GPL via the [documentation](https://sbert.net/examples/domain_adaptation/README.html#gpl-generative-pseudo-labeling), but have never used it myself. As a result, I'm not quite sure how challenging the implementation here would be. That said, for some of the components (e.g. hard negatives mining) it would be useful to have helper methods. And I agree that query generation with T5 is a bit dated nowadays, and Reranking should have some more flexibility as well. I'm also open to PRs for e.g. Negative Mining here, and for PRs on training examples in the Unsupervised Learning code (and the documentation): https://github.com/UKPLab/sentence-transformers/tree/master/examples/unsupervised_learning
- [ ] I'm definitely interested in this, too. I think an integration with `mergekit` for easy merging of Sentence Transformer models would be valuable: https://github.com/arcee-ai/mergekit/issues/286
- [ ] I remember looking into an integration with the LLM2Vec models, but I don't remember what the bottleneck was at the time. I also was not extremely impressed with the performance of their models compared to their size. Perhaps it was related to PEFT, which is somewhat poorly supported in Sentence Transformers at the moment. I'm also looking to improve that support.

Great suggestions!

- Tom Aarsen

**Comment by dmacko232** [comment]
@tomaarsen Thanks for your reply!

I will consider contributing with PRs but I do not have any capacity for some time (month and half). That's why I suggested these features :).

For GPL, I think it could be just decomposed into couple helper functions/examples
- generate positive queries using T5 (or some other generative LLM) 
- mine hard negative candidates using Bi-Encoders
- pseudo-label hard negatives candidates using Cross-Encoders or generative LLMs
- optionally: filter hard negatives candidates (based on the pseudo-labels)
- examples of fine-tuning with pseudo-labeled hard negative candidates but also without the labels (just filtering out the false negatives)

**Comment by tomaarsen** [comment]
On that topic, I'll be picking up the hard-negative mining with bi-encoders (i.e. the SentenceTransformer class) in the short-term, with the possibility to use CrossEncoders for rescoring (as it's convenient to have it all in one function for those who like that). I'll tag you both in the eventual PR for it.

- Tom Aarsen
